// @mysten/walrus/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  MAINNET_WALRUS_PACKAGE_CONFIG: () => import_constants.MAINNET_WALRUS_PACKAGE_CONFIG,
  TESTNET_WALRUS_PACKAGE_CONFIG: () => import_constants.TESTNET_WALRUS_PACKAGE_CONFIG,
  WalrusBlob: () => import_blob.WalrusBlob,
  WalrusClient: () => import_client.WalrusClient,
  WalrusFile: () => import_file.WalrusFile,
  blobIdFromInt: () => import_bcs.blobIdFromInt,
  blobIdToInt: () => import_bcs.blobIdToInt,
  encodeQuilt: () => import_quilts.encodeQuilt,
  walrus: () => import_client.walrus
});
module.exports = __toCommonJS(index_exports);
var import_client = require("./client.js");
var import_constants = require("./constants.js");
__reExport(index_exports, require("./storage-node/error.js"), module.exports);
__reExport(index_exports, require("./error.js"), module.exports);
var import_quilts = require("./utils/quilts.js");
var import_bcs = require("./utils/bcs.js");
var import_file = require("./files/file.js");
var import_blob = require("./files/blob.js");
