// @mysten/walrus/dist/esm/error.js
class WalrusClientError extends Error {
}
class RetryableWalrusClientError extends WalrusClientError {
}
class NoBlobStatusReceivedError extends WalrusClientError {
}
class NoVerifiedBlobStatusReceivedError extends WalrusClientError {
}
class NoBlobMetadataReceivedError extends RetryableWalrusClientError {
}
class NotEnoughSliversReceivedError extends RetryableWalrusClientError {
}
class NotEnoughBlobConfirmationsError extends RetryableWalrusClientError {
}
class BehindCurrentEpochError extends RetryableWalrusClientError {
}
class BlobNotCertifiedError extends RetryableWalrusClientError {
}
class InconsistentBlobError extends WalrusClientError {
}
class BlobBlockedError extends Error {
}
export {
  BehindCurrentEpochError,
  BlobBlockedError,
  BlobNotCertifiedError,
  InconsistentBlobError,
  NoBlobMetadataReceivedError,
  NoBlobStatusReceivedError,
  NoVerifiedBlobStatusReceivedError,
  NotEnoughBlobConfirmationsError,
  NotEnoughSliversReceivedError,
  RetryableWalrusClientError,
  WalrusClientError
};
