// @mysten/walrus/dist/cjs/contracts/walrus/messages.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var messages_exports = {};
__export(messages_exports, {
  BlobPersistenceType: () => BlobPersistenceType,
  CertifiedBlobMessage: () => CertifiedBlobMessage,
  CertifiedInvalidBlobId: () => CertifiedInvalidBlobId,
  CertifiedMessage: () => CertifiedMessage,
  DenyListBlobDeleted: () => DenyListBlobDeleted,
  DenyListUpdateMessage: () => DenyListUpdateMessage,
  ProofOfPossessionMessage: () => ProofOfPossessionMessage,
  ProtocolVersionMessage: () => ProtocolVersionMessage
});
module.exports = __toCommonJS(messages_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
const $moduleName = "@local-pkg/walrus::messages";
const ProofOfPossessionMessage = new import_utils.MoveStruct({
  name: `${$moduleName}::ProofOfPossessionMessage`,
  fields: {
    intent_type: import_bcs.bcs.u8(),
    intent_version: import_bcs.bcs.u8(),
    intent_app: import_bcs.bcs.u8(),
    epoch: import_bcs.bcs.u32(),
    sui_address: import_bcs.bcs.Address,
    bls_key: import_bcs.bcs.vector(import_bcs.bcs.u8())
  }
});
const CertifiedMessage = new import_utils.MoveStruct({
  name: `${$moduleName}::CertifiedMessage`,
  fields: {
    intent_type: import_bcs.bcs.u8(),
    intent_version: import_bcs.bcs.u8(),
    cert_epoch: import_bcs.bcs.u32(),
    message: import_bcs.bcs.vector(import_bcs.bcs.u8()),
    stake_support: import_bcs.bcs.u16()
  }
});
const BlobPersistenceType = new import_utils.MoveEnum({
  name: `${$moduleName}::BlobPersistenceType`,
  fields: {
    Permanent: null,
    Deletable: new import_utils.MoveStruct({
      name: `BlobPersistenceType.Deletable`,
      fields: {
        object_id: import_bcs.bcs.Address
      }
    })
  }
});
const CertifiedBlobMessage = new import_utils.MoveStruct({
  name: `${$moduleName}::CertifiedBlobMessage`,
  fields: {
    blob_id: import_bcs.bcs.u256(),
    blob_persistence_type: BlobPersistenceType
  }
});
const CertifiedInvalidBlobId = new import_utils.MoveStruct({
  name: `${$moduleName}::CertifiedInvalidBlobId`,
  fields: {
    blob_id: import_bcs.bcs.u256()
  }
});
const ProtocolVersionMessage = new import_utils.MoveStruct({
  name: `${$moduleName}::ProtocolVersionMessage`,
  fields: {
    start_epoch: import_bcs.bcs.u32(),
    protocol_version: import_bcs.bcs.u64()
  }
});
const DenyListUpdateMessage = new import_utils.MoveStruct({
  name: `${$moduleName}::DenyListUpdateMessage`,
  fields: {
    storage_node_id: import_bcs.bcs.Address,
    deny_list_sequence_number: import_bcs.bcs.u64(),
    deny_list_size: import_bcs.bcs.u64(),
    deny_list_root: import_bcs.bcs.u256()
  }
});
const DenyListBlobDeleted = new import_utils.MoveStruct({
  name: `${$moduleName}::DenyListBlobDeleted`,
  fields: {
    blob_id: import_bcs.bcs.u256()
  }
});
