// @mysten/walrus/dist/cjs/contracts/walrus/pool_exchange_rate.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var pool_exchange_rate_exports = {};
__export(pool_exchange_rate_exports, {
  PoolExchangeRate: () => PoolExchangeRate
});
module.exports = __toCommonJS(pool_exchange_rate_exports);
var import_utils = require("../utils/index.js");
var import_bcs = require("@mysten/sui/bcs");
const $moduleName = "@local-pkg/walrus::pool_exchange_rate";
const PoolExchangeRate = new import_utils.MoveEnum({
  name: `${$moduleName}::PoolExchangeRate`,
  fields: {
    Flat: null,
    Variable: new import_utils.MoveStruct({
      name: `PoolExchangeRate.Variable`,
      fields: {
        wal_amount: import_bcs.bcs.u128(),
        share_amount: import_bcs.bcs.u128()
      }
    })
  }
});
