// @mysten/seal/dist/cjs/session-key.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __typeError = (msg) => {
  throw TypeError(msg);
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __accessCheck = (obj, member, msg) => member.has(obj) || __typeError("Cannot " + msg);
var __privateGet = (obj, member, getter) => (__accessCheck(obj, member, "read from private field"), getter ? getter.call(obj) : member.get(obj));
var __privateAdd = (obj, member, value) => member.has(obj) ? __typeError("Cannot add the same private member more than once") : member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
var __privateSet = (obj, member, value, setter) => (__accessCheck(obj, member, "write to private field"), setter ? setter.call(obj, value) : member.set(obj, value), value);
var session_key_exports = {};
__export(session_key_exports, {
  RequestFormat: () => RequestFormat,
  SessionKey: () => SessionKey
});
module.exports = __toCommonJS(session_key_exports);
var import_bcs = require("@mysten/bcs");
var import_bcs2 = require("@mysten/sui/bcs");
var import_ed25519 = require("@mysten/sui/keypairs/ed25519");
var import_utils = require("@mysten/sui/utils");
var import_verify = require("@mysten/sui/verify");
var import_elgamal = require("./elgamal.js");
var import_error = require("./error.js");
var _address, _packageId, _mvrName, _creationTimeMs, _ttlMin, _sessionKey, _personalMessageSignature, _signer, _suiClient;
const RequestFormat = import_bcs2.bcs.struct("RequestFormat", {
  ptb: import_bcs2.bcs.byteVector(),
  encKey: import_bcs2.bcs.byteVector(),
  encVerificationKey: import_bcs2.bcs.byteVector()
});
const _SessionKey = class _SessionKey {
  constructor({
    address,
    packageId,
    mvrName,
    ttlMin,
    signer,
    suiClient
  }) {
    __privateAdd(this, _address);
    __privateAdd(this, _packageId);
    __privateAdd(this, _mvrName);
    __privateAdd(this, _creationTimeMs);
    __privateAdd(this, _ttlMin);
    __privateAdd(this, _sessionKey);
    __privateAdd(this, _personalMessageSignature);
    __privateAdd(this, _signer);
    __privateAdd(this, _suiClient);
    if (mvrName && !(0, import_utils.isValidNamedPackage)(mvrName)) {
      throw new import_error.UserError(`Invalid package name ${mvrName}`);
    }
    if (!(0, import_utils.isValidSuiObjectId)(packageId) || !(0, import_utils.isValidSuiAddress)(address)) {
      throw new import_error.UserError(`Invalid package ID ${packageId} or address ${address}`);
    }
    if (ttlMin > 30 || ttlMin < 1) {
      throw new import_error.UserError(`Invalid TTL ${ttlMin}, must be between 1 and 30`);
    }
    if (signer && signer.getPublicKey().toSuiAddress() !== address) {
      throw new import_error.UserError("Signer address does not match session key address");
    }
    __privateSet(this, _address, address);
    __privateSet(this, _packageId, packageId);
    __privateSet(this, _mvrName, mvrName);
    __privateSet(this, _creationTimeMs, Date.now());
    __privateSet(this, _ttlMin, ttlMin);
    __privateSet(this, _sessionKey, import_ed25519.Ed25519Keypair.generate());
    __privateSet(this, _signer, signer);
    __privateSet(this, _suiClient, suiClient);
  }
  /**
   * Create a new SessionKey instance.
   * @param address - The address of the user.
   * @param packageId - The ID of the package.
   * @param mvrName - Optional. The name of the MVR, if there is one.
   * @param ttlMin - The TTL in minutes.
   * @param signer - Optional. The signer instance, e.g. EnokiSigner.
   * @param suiClient - The Sui client.
   * @returns A new SessionKey instance.
   */
  static async create({
    address,
    packageId,
    mvrName,
    ttlMin,
    signer,
    suiClient
  }) {
    const packageObj = await suiClient.core.getObject({ objectId: packageId });
    if (String(packageObj.object.version) !== "1") {
      throw new import_error.InvalidPackageError(`Package ${packageId} is not the first version`);
    }
    return new _SessionKey({
      address,
      packageId,
      mvrName,
      ttlMin,
      signer,
      suiClient
    });
  }
  isExpired() {
    return __privateGet(this, _creationTimeMs) + __privateGet(this, _ttlMin) * 60 * 1e3 - 1e4 < Date.now();
  }
  getAddress() {
    return __privateGet(this, _address);
  }
  getPackageName() {
    if (__privateGet(this, _mvrName)) {
      return __privateGet(this, _mvrName);
    }
    return __privateGet(this, _packageId);
  }
  getPackageId() {
    return __privateGet(this, _packageId);
  }
  getPersonalMessage() {
    const creationTimeUtc = new Date(__privateGet(this, _creationTimeMs)).toISOString().slice(0, 19).replace("T", " ") + " UTC";
    const message = `Accessing keys of package ${this.getPackageName()} for ${__privateGet(this, _ttlMin)} mins from ${creationTimeUtc}, session key ${(0, import_bcs.toBase64)(__privateGet(this, _sessionKey).getPublicKey().toRawBytes())}`;
    return new TextEncoder().encode(message);
  }
  async setPersonalMessageSignature(personalMessageSignature) {
    if (!__privateGet(this, _personalMessageSignature)) {
      try {
        await (0, import_verify.verifyPersonalMessageSignature)(this.getPersonalMessage(), personalMessageSignature, {
          address: __privateGet(this, _address),
          client: __privateGet(this, _suiClient)
        });
        __privateSet(this, _personalMessageSignature, personalMessageSignature);
      } catch {
        throw new import_error.InvalidPersonalMessageSignatureError("Not valid");
      }
    }
  }
  async getCertificate() {
    if (!__privateGet(this, _personalMessageSignature)) {
      if (__privateGet(this, _signer)) {
        const { signature } = await __privateGet(this, _signer).signPersonalMessage(this.getPersonalMessage());
        __privateSet(this, _personalMessageSignature, signature);
      } else {
        throw new import_error.InvalidPersonalMessageSignatureError("Personal message signature is not set");
      }
    }
    return {
      user: __privateGet(this, _address),
      session_vk: (0, import_bcs.toBase64)(__privateGet(this, _sessionKey).getPublicKey().toRawBytes()),
      creation_time: __privateGet(this, _creationTimeMs),
      ttl_min: __privateGet(this, _ttlMin),
      signature: __privateGet(this, _personalMessageSignature),
      mvr_name: __privateGet(this, _mvrName)
    };
  }
  /**
   * Create request params for the given transaction bytes.
   * @param txBytes - The transaction bytes.
   * @returns The request params containing the ephemeral secret key,
   * its public key and its verification key.
   */
  async createRequestParams(txBytes) {
    if (this.isExpired()) {
      throw new import_error.ExpiredSessionKeyError();
    }
    const encKey = (0, import_elgamal.generateSecretKey)();
    const encKeyPk = (0, import_elgamal.toPublicKey)(encKey);
    const encVerificationKey = (0, import_elgamal.toVerificationKey)(encKey);
    const msgToSign = RequestFormat.serialize({
      ptb: txBytes.slice(1),
      encKey: encKeyPk,
      encVerificationKey
    }).toBytes();
    return {
      encKey,
      encKeyPk,
      encVerificationKey,
      requestSignature: (0, import_bcs.toBase64)(await __privateGet(this, _sessionKey).sign(msgToSign))
    };
  }
  /**
   * Export the Session Key object from the instance. Store the object in IndexedDB to persist.
   */
  export() {
    const obj = {
      address: __privateGet(this, _address),
      packageId: __privateGet(this, _packageId),
      mvrName: __privateGet(this, _mvrName),
      creationTimeMs: __privateGet(this, _creationTimeMs),
      ttlMin: __privateGet(this, _ttlMin),
      personalMessageSignature: __privateGet(this, _personalMessageSignature),
      sessionKey: __privateGet(this, _sessionKey).getSecretKey()
      // bech32 encoded string
    };
    Object.defineProperty(obj, "toJSON", {
      enumerable: false,
      value: () => {
        throw new Error("This object is not serializable");
      }
    });
    return obj;
  }
  /**
   * Restore a SessionKey instance for the given object.
   * @returns A new SessionKey instance with restored state
   */
  static import(data, suiClient, signer) {
    const instance = new _SessionKey({
      address: data.address,
      packageId: data.packageId,
      mvrName: data.mvrName,
      ttlMin: data.ttlMin,
      signer,
      suiClient
    });
    __privateSet(instance, _creationTimeMs, data.creationTimeMs);
    __privateSet(instance, _sessionKey, import_ed25519.Ed25519Keypair.fromSecretKey(data.sessionKey));
    __privateSet(instance, _personalMessageSignature, data.personalMessageSignature);
    if (instance.isExpired()) {
      throw new import_error.ExpiredSessionKeyError();
    }
    return instance;
  }
};
_address = new WeakMap();
_packageId = new WeakMap();
_mvrName = new WeakMap();
_creationTimeMs = new WeakMap();
_ttlMin = new WeakMap();
_sessionKey = new WeakMap();
_personalMessageSignature = new WeakMap();
_signer = new WeakMap();
_suiClient = new WeakMap();
let SessionKey = _SessionKey;
