// @mysten/seal/dist/cjs/utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  MAX_U8: () => MAX_U8,
  Version: () => Version,
  allEqual: () => allEqual,
  count: () => count,
  createFullId: () => createFullId,
  equals: () => equals,
  flatten: () => flatten,
  hasDuplicates: () => hasDuplicates,
  xor: () => xor,
  xorUnchecked: () => xorUnchecked
});
module.exports = __toCommonJS(utils_exports);
var import_bcs = require("@mysten/bcs");
var import_utils = require("@mysten/sui/utils");
var import_error = require("./error.js");
const MAX_U8 = 255;
function xor(a, b) {
  if (a.length !== b.length) {
    throw new Error("Invalid input");
  }
  return xorUnchecked(a, b);
}
function xorUnchecked(a, b) {
  return a.map((ai, i) => ai ^ b[i]);
}
function createFullId(packageId, innerId) {
  if (!(0, import_utils.isValidSuiObjectId)(packageId)) {
    throw new import_error.UserError(`Invalid package ID ${packageId}`);
  }
  const fullId = flatten([(0, import_bcs.fromHex)(packageId), (0, import_bcs.fromHex)(innerId)]);
  return (0, import_bcs.toHex)(fullId);
}
function flatten(arrays) {
  const length = arrays.reduce((sum, arr) => sum + arr.length, 0);
  const result = new Uint8Array(length);
  arrays.reduce((offset, array) => {
    result.set(array, offset);
    return offset + array.length;
  }, 0);
  return result;
}
function count(array, value) {
  return array.reduce((count2, item) => item === value ? count2 + 1 : count2, 0);
}
function hasDuplicates(array) {
  return new Set(array).size !== array.length;
}
function allEqual(array) {
  if (array.length === 0) {
    return true;
  }
  return array.every((item) => item === array[0]);
}
function equals(a, b) {
  if (a.length !== b.length) {
    return false;
  }
  return a.every((ai, i) => ai === b[i]);
}
class Version {
  constructor(version) {
    const parts = version.split(".").map(Number);
    if (parts.length !== 3 || parts.some((part) => isNaN(part) || !Number.isInteger(part) || part < 0)) {
      throw new import_error.UserError(`Invalid version format: ${version}`);
    }
    this.major = parts[0];
    this.minor = parts[1];
    this.patch = parts[2];
  }
  // Compare this version with another version. True if this version is older than the other version.
  older_than(other) {
    if (this.major !== other.major) {
      return this.major < other.major;
    } else if (this.minor !== other.minor) {
      return this.minor < other.minor;
    }
    return this.patch < other.patch;
  }
}
